### PR TITLE
Port GTSFM's DSF track generator to C++

### PR DESCRIPTION
**Pull request by johnwlambert**
DSF track generation speed is prohibitively slow in Python for certain front-ends, so the code is ported to C++ for efficiency.

Corresponding GTSFM code can be found at https://github.com/borglab/gtsfm/blob/master/gtsfm/data_association/dsf_tracks_estimator.py.

### CODE REVIEW

**Comment by varunagrawal** [comment]
Update to use C++11 please? External folks still expect C++11 conformity.

File: gtsam/sfm/DsfTrackGenerator.h [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+/**
+ * @file DsfTrackGenerator.h
+ * @date July 2022
+ * @author John Lambert
+ * @brief Identifies connected components in the keypoint matched graph.
+ */
+
+#pragma once
+#include <pybind11/stl.h>
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:20]: do we need this?

File: gtsam/sfm/DsfTrackGenerator.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+
+struct Keypoints
+{
+  KeypointCoordinates coordinates;
+  // typedef'd for Eigen::VectorXd
+  boost::optional<gtsam::Vector> scales;
+  boost::optional<gtsam::Vector> responses;
+
+  Keypoints(const gtsam::KeypointCoordinates& coordinates): coordinates(coordinates) {}; // boost::none
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:49]: why do we need these constructors for these structs? same about the other structs in this file.

File: gtsam/sfm/DsfTrackGenerator.h [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+using KeypointsList = std::vector<Keypoints>;
+using KeypointsVector = std::vector<Keypoints>; // TODO(johnwlambert): prefer KeypointsSet?
+using MatchIndicesMap = std::map<ImagePair, CorrespondenceIndices>;
+
+
+// @param camera index
+// @param 2d measurement
+// Implemented as named tuple, instead of std::pair (like SfmMeasurement in SfmTrack.h)
+struct NamedSfmMeasurement
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:60]: Does "Named" make a difference to the user? If not we could just call it SfmMeasurement.

File: gtsam/sfm/DsfTrackGenerator.h [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+{
+  size_t i;
+  gtsam::Point2 uv;
+
+  NamedSfmMeasurement(size_t i, gtsam::Point2 uv) : i(i), uv(uv) {}
+};
+
+
+class SfmTrack2d
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:69]: Missing documentation.

File: gtsam/sfm/DsfTrackGenerator.h [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+{
+
+  public:
+    DsfTrackGenerator() {}
+
+    // Creates a list of tracks from 2d point correspondences.
+    //
+    // Creates a disjoint-set forest (DSF) and 2d tracks from pairwise matches. We create a singleton for union-find
+    // set elements from camera index of a detection and the index of that detection in that camera's keypoint list,
+    // i.e. (i,k).
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:107]: this looks like its not formatted according to gtsam conventions

File: gtsam/sfm/DsfTrackGenerator.h [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    std::vector<SfmTrack2d> generate_tracks_from_pairwise_matches(
+        const MatchIndicesMap& matches_dict,
+        const KeypointsList& keypoints_list)
+    {
+      std::vector<SfmTrack2d> track_2d_list;
+
+      std::cout << "[SfmTrack2d] Starting Union-Find..." << std::endl;
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:114]: the union find is an implementation detail I think, no need to log that.. 
Maybe this function does not need any std::cout statements. You could leave the one that prints the number of tracks if you'd like.

File: gtsam/sfm/DsfTrackGenerator.h [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+          // Camera index is represented by i, and measurement index is represented by k.
+          size_t i = index_pair.i();
+          size_t k = index_pair.j();
+          // Add measurement to this track.
+          track_2d.addMeasurement(NamedSfmMeasurement(i, keypoints_list[i].coordinates.row(k)));
+        }
+
+        // Skip erroneous track that had repeated measurements within the same image.
+        // This is an expected result from an incorrect correspondence slipping through.
+        if (track_2d.validate_unique_cameras())
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:153]: nit: has_unique_cameras() seems like a better name.

File: gtsam/sfm/DsfTrackGenerator.h [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        {
+          track_2d_list.emplace_back(track_2d);
+        } else {
+          erroneous_track_count += 1;
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:157]: nit: ++

File: gtsam/sfm/DsfTrackGenerator.h [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        }
+      }
+
+      double erroneous_track_pct = key_sets.size() > 0 ? erroneous_track_count / key_sets.size() * 100 : std::nan("1");
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:161]: division by two ints will give an int, so 44/90 => 0

Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:161]: handle the keys_sets.empty() case at the beginning of the function? then you would not need the NaNs

File: gtsam/sfm/DsfTrackGenerator.h [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+  // typedef'd for Eigen::VectorXd
+  boost::optional<gtsam::Vector> scales;
+  boost::optional<gtsam::Vector> responses;
+
+  Keypoints(const gtsam::KeypointCoordinates& coordinates): coordinates(coordinates) {}; // boost::none
+};
+
+using KeypointsList = std::vector<Keypoints>;
+using KeypointsVector = std::vector<Keypoints>; // TODO(johnwlambert): prefer KeypointsSet?
+using MatchIndicesMap = std::map<ImagePair, CorrespondenceIndices>;
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:54]: why do we need a map? is this for gtsfm's convenience?

Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:54]: It's used for 1:1 parity between the inputs to the Python version of this code, and to the C++ version of this code.

Can't think of a cleaner way to provide the input arguments without the map.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 3 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:60]: It would conflict with [the existing `SfmMeasurement`](https://github.com/borglab/gtsam/blob/develop/gtsam/sfm/SfmTrack.h#L32) here, and I really want the unambiguous named attributes, for parity with the Python side.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 4 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:69]: Added, thanks.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 8 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:157]: Updated, thanks.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 9 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:161]: Thanks, updated.

Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:161]: Thanks, updated.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 7 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:153]: Updated, thanks.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 5 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:107]: Cleaned up the line length -- let me know if this looks appropriate now.

Comment (akshay-krishnan) [reply gtsam/sfm/DsfTrackGenerator.h]: There are other formatting issues as well. For example, `{ ` should start on the same line. 
We use the google style formatting, can you try enabling that in your IDE for C++ and re-formatting?

File: gtsam/sfm/DsfTrackGenerator.h (hunk 3 above)
Comment (akshay-krishnan) [reply gtsam/sfm/DsfTrackGenerator.h:60]: why cant we use the existing tuple?

**Comment by johnwlambert** [comment]
Hi @ProfFan @varunagrawal any idea why the Python release builds would succeed, but the Python debug builds would fail in CI?

**Comment by ProfFan** [comment]
@johnwlambert Probably the same issue as #1244

File: gtsam/sfm/sfm.i [hunk 11]
```diff
@@ -4,6 +4,60 @@
...
+#include <gtsam/sfm/DsfTrackGenerator.h>
+
+class MatchIndicesMap {
+  MatchIndicesMap();
+  MatchIndicesMap(const gtsam::MatchIndicesMap& other);
+
+  size_t size() const;
+  bool empty() const;
+  void clear();
+  gtsam::CorrespondenceIndices at(const pair<size_t, size_t>& keypair) const;
```
Comment (akshay-krishnan) [inline gtsam/sfm/sfm.i:16]: has CorrespondenceIndices been wrapped?

File: gtsam/sfm/sfm.i [hunk 12]
```diff
@@ -4,6 +4,60 @@
...
+  KeypointsList(const gtsam::KeypointsList& other);
+  void push_back(const gtsam::Keypoints& keypoints);
+  size_t size() const;
+  bool empty() const;
+  void clear();
+  gtsam::Keypoints at(const size_t& index) const;
+};
+
+
+class Keypoints
```
Comment (akshay-krishnan) [inline gtsam/sfm/sfm.i:31]: this should go before KeypointsList

File: gtsam/sfm/sfm.i [hunk 13]
```diff
@@ -4,6 +4,60 @@
...
+{
+  Keypoints(const gtsam::KeypointCoordinates& coordinates);
+  gtsam::KeypointCoordinates coordinates;
```
Comment (akshay-krishnan) [inline gtsam/sfm/sfm.i:34]: has KeypointCoordinates been wrapped?

File: gtsam/sfm/DsfTrackGenerator.h [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+  KeypointCoordinates coordinates;
+  // typedef'd for Eigen::VectorXd
+  boost::optional<gtsam::Vector> scales;
+  boost::optional<gtsam::Vector> responses;
+
+  Keypoints(const gtsam::KeypointCoordinates& coordinates): coordinates(coordinates) {}; // boost::none
+};
+
+using KeypointsList = std::vector<Keypoints>;
+using KeypointsVector = std::vector<Keypoints>; // TODO(johnwlambert): prefer KeypointsSet?
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:51]: do we need both?

File: gtsam/sfm/DsfTrackGenerator.h [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+ */
+class SfmTrack2d {
+ private:
+  std::vector<NamedSfmMeasurement> measurements_;
+
+ public:
+  void addMeasurement(const NamedSfmMeasurement &m) {
+    measurements_.emplace_back(m);
+  }
+  std::vector<NamedSfmMeasurement> measurements() {return measurements_; }
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:79]: how do we use the camera index in the returned results?

File: gtsam/sfm/DsfTrackGenerator.h [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+
+  // @brief Validates the track by checking that no two measurements are from the same camera.
+  //
+  // returns boolean result of the validation.
+  bool has_unique_cameras()
```
[the remark this answers is not in the document]
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:84]: Updated now, thanks.

File: gtsam/sfm/sfm.i (hunk 12 above)
Comment (johnwlambert) [reply gtsam/sfm/sfm.i:31]: Thanks, agreed. Re-ordered now.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 14 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:51]: Wasn't sure what our current convention for this is. I think `KeypointsVector` could create some confusion since it's a vector of keypoint vectors (2d coordinates). So I'd prefer `KeypointsList` or `KeypointsSet`.

File: gtsam/sfm/DsfTrackGenerator.h [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+
+using KeypointsList = std::vector<Keypoints>;
+using KeypointsVector = std::vector<Keypoints>; // TODO(johnwlambert): prefer KeypointsSet?
+using MatchIndicesMap = std::map<ImagePair, CorrespondenceIndices>;
+
+
+// @param camera index
+// @param 2d measurement
+// Implemented as named tuple, instead of std::pair (like SfmMeasurement in SfmTrack.h)
+struct NamedSfmMeasurement {
```
[the remark this answers is not in the document]
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:58]: The only issue with renaming `i` to `cameraIdx` is that we won't be able to use the GTSAM SfmTrack2d and its GTSFM equivalent identically (to keep the same name for both).

Would `IndexedSfmMeasurement` imply we could index into it?

File: gtsam/sfm/DsfTrackGenerator.h [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <optional>
+#include <vector>
+
+namespace gtsam {
+
+typedef DSFMap<IndexPair> DSFMapIndexPair;
+typedef std::pair<size_t, size_t> ImagePair;
```
[the remark this answers is not in the document]
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h]: Good point, removed now.

**Comment by varunagrawal** [comment]
It's not related to #1244. The issue was something to do with Github Actions which has been resolved so you can pull in the latest develop.

**Comment by ProfFan** [comment]
Ping @johnwlambert

File: gtsam/sfm/DsfTrackGenerator.h [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+typedef Eigen::MatrixX2i CorrespondenceIndices; // N x 2 array
+
+//struct Keypoints;
+using KeypointCoordinates = Eigen::MatrixX2d;
```
[the remark this answers is not in the document]
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:37]: Thanks. I see what you mean but I think it sounds a bit strange. @dellaert any thoughts on this one?

File: gtsam/sfm/DsfTrackGenerator.h (hunk 14 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:50]: @dellaert @akshay-krishnan any votes on the naming between KeypointsVector vs KeypointsList?

File: gtsam/sfm/sfm.i (hunk 11 above)
Comment (johnwlambert) [reply gtsam/sfm/sfm.i:16]: No need to wrap it, since it's just a typedef on Eigen, so the conversion should be implicit

```c++
typedef Eigen::MatrixX2i CorrespondenceIndices; // N x 2 array
```

File: gtsam/sfm/DsfTrackGenerator.h (hunk 15 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:79]: One example -- we unpack it here:
```python
    def extract_measurements(
        self, track: SfmTrack2d
         ...
    )
         for i, uv in track.measurements:
```

https://github.com/borglab/gtsfm/blob/master/gtsfm/data_association/point3d_initializer.py#L352

**Comment by johnwlambert** [comment]
@akshay-krishnan @dellaert the CI passes now, so once we resolve the style questions above, this will be good to merge

File: gtsam/sfm/sfm.i (hunk 13 above)
Comment (johnwlambert) [reply gtsam/sfm/sfm.i]: No need to wrap it, since it's just a typedef on Eigen, with implicit conversion

```c++
using KeypointCoordinates = Eigen::MatrixX2d;
```

File: gtsam/sfm/DsfTrackGenerator.h [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+  }
+
+  /// The measurement at index `idx`
+  NamedSfmMeasurement measurement(size_t idx) const { return measurements_[idx]; }
+
+  // Return all measurements in the track.
+  std::vector<NamedSfmMeasurement> measurements() {return measurements_; }
+
+  /// Total number of measurements in this track.
+  size_t numberMeasurements() const { return measurements_.size(); }
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:94]: nit: I think Frank started using nrMeasurements() in another class.. but up to you.

File: gtsam/sfm/DsfTrackGenerator.h [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+  // i.e. (i,k).
+  // @param Map from (i1,i2) image pair indices to (K,2) matrix, for K
+  //        correspondence indices, from each image.
+  // @param Length-N list of keypoints, for N images/cameras.
+  std::vector<SfmTrack2d> generate_tracks_from_pairwise_matches(
+    const MatchIndicesMap& matches_dict,
+    const KeypointsList& keypoints_list) {
+    std::vector<SfmTrack2d> track_2d_list;
+
+    std::cout << "[SfmTrack2d] Starting Union-Find..." << std::endl;
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:139]: Do we need this?

Comment (akshay-krishnan) [reply gtsam/sfm/DsfTrackGenerator.h:139]: Maybe only do it if a verbose flag is set ?

File: gtsam/sfm/DsfTrackGenerator.h [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+      {
+        // Measurement indices are found in a single matrix row, as (k1,k2).
+        size_t k1 = corr_idxs(k, 0);
+        size_t k2 = corr_idxs(k, 1);
+        // Unique keys for the DSF are (i,k), representing keypoint index in an image.
+        dsf.merge(IndexPair(i1, k1), IndexPair(i2, k2));
+      }
+    }
+
+    std::cout << "[SfmTrack2d] Union-Find Complete" << std::endl;
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:160]: do we need this?

File: gtsam/sfm/DsfTrackGenerator.h [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+      } else {
+        erroneous_track_count++;
+      }
+    }
+
+    double erroneous_track_pct = static_cast<float>(erroneous_track_count)
+                                 / static_cast<float>(key_sets.size()) * 100;
+    // TODO(johnwlambert): restrict decimal places to 2 decimals.
+    std::cout << "DSF Union-Find: " << erroneous_track_pct;
+    std::cout << "% of tracks discarded from multiple obs. in a single image." << std::endl;
```
Comment (akshay-krishnan) [inline gtsam/sfm/DsfTrackGenerator.h:199]: same comment, do we need this is a low-level API ?

File: python/gtsam/tests/test_DsfTrackGenerator.py [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+        assert tracks[1].measurements()[2].i == 2
+        assert tracks[1].numberMeasurements() == 3
+
+        # Verify track 2.
+        assert np.allclose(tracks[2].measurements()[0].uv, np.array([90.0, 100.0]))
+        assert np.allclose(tracks[2].measurements()[1].uv, np.array([110.0, 120.0]))
+        assert tracks[2].measurements()[0].i == 1
+        assert tracks[2].measurements()[1].i == 2
+        assert tracks[2].numberMeasurements() == 2
+
```
Comment (akshay-krishnan) [inline python/gtsam/tests/test_DsfTrackGenerator.py:68]: I think the tests are insufficient but can be covered in a different PR. Correct me if I am wrong, but this is a naive DSF class right? So if there are tracks with multiple points in the same image, they are still considered?

**Review by akshay-krishnan — APPROVED** [review]
Conditional approval, pending unresolved comments.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 20 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:94]: I think we changed it to ‘numberMeasurements’ before merging in SfmTrack.h (https://github.com/borglab/gtsam/blob/develop/gtsam/sfm/SfmTrack.h#L71)

File: gtsam/sfm/DsfTrackGenerator.h [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+using KeypointCoordinates = Eigen::MatrixX2d;
+
+
+struct Keypoints {
+  KeypointCoordinates coordinates;
+  // typedef'd for Eigen::VectorXd
+  boost::optional<gtsam::Vector> scales;
+  boost::optional<gtsam::Vector> responses;
+
+  Keypoints(const gtsam::KeypointCoordinates& coordinates): coordinates(coordinates) {}; // boost::none
```
[the remark this answers is not in the document]
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:46]: I don’t think we would have a way to initialize this type on the Python side without defining this constructor right? Although on the c++ side, a struct has the brace initializer syntax that comes for free, I think we need to define a constructor in the .h file to use it in .i, for python use.

File: gtsam/sfm/DsfTrackGenerator.h [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+
+
+// @param camera index
+// @param 2d measurement
+// Implemented as named tuple, instead of std::pair (like SfmMeasurement in SfmTrack.h)
+struct NamedSfmMeasurement {
+  size_t i;
+  gtsam::Point2 uv;
+
+  NamedSfmMeasurement(size_t i, gtsam::Point2 uv) : i(i), uv(uv) {}
```
[the remark this answers is not in the document]
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:62]: Don’t we need it if we want to expose this type on the Python side?

[the remark this answers is not in the document]
Comment (akshay-krishnan) [reply gtsam/sfm/DsfTrackGenerator.h:62]: Im not sure, I wonder if pybind adds a default constructor? 
Otherwise you should be able to do 
```
measurement = NamedSfMMeasurement()
measurement.i = 0
measurement.uv = (0, 1)
```

File: gtsam/sfm/DsfTrackGenerator.h (hunk 14 above)
Comment (akshay-krishnan) [reply gtsam/sfm/DsfTrackGenerator.h:50]: either works for me, I think these names are getting pretty confusing anyway. Ideally APIs using std::vector<Keypoints> should be implicitly converted to a list in python without needing a name for it :)

File: gtsam/sfm/DsfTrackGenerator.h (hunk 20 above)
Comment (akshay-krishnan) [reply gtsam/sfm/DsfTrackGenerator.h:94]: oh i see, thanks!

File: gtsam/sfm/DsfTrackGenerator.h (hunk 25 above)
[the remark this answers is not in the document]
Comment (akshay-krishnan) [reply gtsam/sfm/DsfTrackGenerator.h:46]: have you looked at the constructors in python without defining this? related to another comment I just responded to.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 14 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:50]: Agreed. @ProfFan @dellaert is the use of stl.h's implicit conversion something on the roadmap or is it too bug-prone?

File: gtsam/sfm/DsfTrackGenerator.h (hunk 26 above)
[the remark this answers is not in the document]
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:62]: From my recollection the constructor would have to be explicit in the .i file to be registered. I think the keyword arg constructor is a bit cleaner (1 line vs 3 lines) : - )

File: gtsam/sfm/DsfTrackGenerator.h [hunk 27]
```diff
@@ -0,0 +1,207 @@
...
+namespace gtsam {
+
+typedef DSFMap<IndexPair> DSFMapIndexPair;
+typedef Eigen::MatrixX2i CorrespondenceIndices; // N x 2 array
+
+//struct Keypoints;
+using KeypointCoordinates = Eigen::MatrixX2d;
+
+
+struct Keypoints {
```
[the remark this answers is not in the document]
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:40]: Thanks, added some documentation now.

**Comment by akshay-krishnan** [comment]
@johnwlambert ping about this, we wanted to get this merged and build a new release.

File: gtsam/sfm/DsfTrackGenerator.h [hunk 28]
```diff
@@ -0,0 +1,217 @@
...
+  // Note: gtsam::Vector is typedef'd for Eigen::VectorXd.
+  boost::optional<gtsam::Vector> scales;
+
+  /// Optional confidences/responses for each detection, of shape N.
+  boost::optional<gtsam::Vector> responses;
+
+  Keypoints(const gtsam::KeypointCoordinates& coordinates): coordinates(coordinates) {}; // boost::none
+};
+
+using KeypointsList = std::vector<Keypoints>;
```
Comment (dellaert) [inline gtsam/sfm/DsfTrackGenerator.h:60]: KeypointsVector

File: gtsam/sfm/DsfTrackGenerator.h [hunk 29]
```diff
@@ -0,0 +1,217 @@
...
+#include <gtsam/geometry/Point2.h>
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <optional>
+#include <vector>
+
+namespace gtsam {
+
```
Comment (dellaert) [inline gtsam/sfm/DsfTrackGenerator.h:32]: Can we push all these in a namespace?

File: gtsam/sfm/DsfTrackGenerator.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (dellaert) [reply gtsam/sfm/DsfTrackGenerator.h:68]: Just use SfmMeasurement, nuke this struct.

In python, you should be able to just say:

```
i, uv = measurement
```

File: gtsam/sfm/DsfTrackGenerator.h (hunk 28 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h]: Thanks, updated KeypointsList to KeypointsVector.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 1 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h]: Removed, thanks.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 23 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h]: Thanks, added a `verbose` flag to the method that defaults to false that controls the logging verbosity.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 22 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h]: Thanks, added a `verbose` flag to the method that defaults to false that controls the logging verbosity.

For very large sequences, this logging can provide helpful visibility during extended execution regarding the current execution state.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 21 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h]: Thanks, added a `verbose` flag to the method that defaults to false that controls the logging verbosity.

File: gtsam/sfm/DsfTrackGenerator.h (hunk 29 above)
Comment (johnwlambert) [reply gtsam/sfm/DsfTrackGenerator.h:32]: Sure, moved them to a `gtsam.gtsfm` namespace now.

**Comment by johnwlambert** [comment]
Superseded by https://github.com/borglab/gtsam/pull/1312
